### PR TITLE
Add dropout hook to draw one weight mask per forward pass

### LINKED ISSUE #500: Fix torch Dropout sampling
`nn.Dropout` in train mode draws an independent mask per sample in the batch. When the representer passes all grid points as a single batch each point gets a different random sub-network in every forward pass. This resulted in a "dotty" plot, while working on #492.
`DropConnectLinear` draws one mask per forward pass, which resulted in a smoother (not "dotty") plot.

### PR DESCRIPTION
**Pull request by MauriceFromm**
## Issue

[Fix torch Dropout sampling](https://github.com/pwhofman/probly/issues/500)

## Motivation and Context

This pr introduces `_install_shared_dropout_hook` in `representer.sampler.torch`, which intercepts the dropout output and applies a single shared binary mask to the output. This matches the behaviour of `DropConnectLinear`, which draws one weight mask per forward pass.

This is applied to all nn.Dropout variants to cover user-supplied models, not just models transformed via `transformation.dropout`.

---

## Public API Changes

-   [ ] No Public API changes
-   [ ] Yes, Public API changes (Details below)

---

## How Has This Been Tested?

---

## Checklist

-   [ ] The changes have been tested locally.
-   [ ] Documentation has been updated (if the public API or usage changes).
-   [ ] A entry has been added to [`CHANGELOG.md`](https://github.com/pwhofman/probly/blob/main/CHANGELOG.md) (if relevant for users).
-   [ ] The code follows the project's [style guidelines](https://github.com/pwhofman/probly/blob/main/.github/CONTRIBUTING.md) and passes minimal code style checks.
-   [ ] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

**Comment by MauriceFromm** [comment]
@pwhofman @Cortys This has been refactored to add a boolean flag in the representer call per your request. Reclassified as feature instead of bug.
A `SHARED_DROPOUT_MASK` GlobalVariable is added to `sampler.common`. The representer can be called via `representer(model, num_samples=x, shared_dropout_mask=True)` to enable this behavior, default is False.

**Review by pwhofman — CHANGES_REQUESTED** [review]
The hook is registered for all currently support DropOut variants, but I don't think it functions correctly for all of them. For example, a Dropout2d or AlphaDropout layer works differently than the regular Dropout and therefore the mask will not be correct. Could you make it work for all of the currently supported DropOut variants? In addition, could you add tests?

**Review by pwhofman — APPROVED** [review]
